### ISSUE TITLE
Proposal: lovr.log

### ISSUE THREAD

**Issue by bjornbytes** [bjornbytes/lovr#177]
This is a proposal to add a logging system to LÖVR.

TLDR;

- Add `lovr.log` callback, it's like `print`.
- C code will call `lovr.log` to log things.
- Lua code can also call `lovr.log` to log things.
- You can override `lovr.log` to render logs in VR or add other logic.
- There is a convention for different log levels.

Goals:

- Expose important information from internal systems, like OpenGL/Vulkan/OpenXR debug callbacks, performance warnings, or deprecation notices.
- Add the ability to intercept logs to render them in VR, filter them, or redirect them.
- Establish a convention for libraries to log things, at different log levels.
- Create a point of indirection that allows Android to work properly without "hacks" (normal printing does not work on Android and we monkeypatch `print` to work right now).

Plan:

- Add a `lovr.log` callback.  The default implementation is in C and prints to stdout on desktop and does android logging on Android.  It accepts multiple arguments, like `print`.
- You can override the callback to do whatever you want, can also save the original implementation into a variable and use it.  `print = lovr.log` might be common.
- C code can call a `lovrLog` function, which will redirect to `lovr.log`.
- Log levels are by convention: Anything can add an allcaps log level as its first argument in square brackets, like this: `lovr.log('[WARN]', 'message')`.  There is nothing special about this, but it could be used by the callback to filter or color the log messages.
  - I think this is better than requiring people to always specify the log level or having separate functions for each level.  But I'm still not sure if it's good.
  - Projects can come up with their own additional conventions (timestamps, log tags)?

Questions:

- Should we just use and override `print` for all this?
- New versions of Lua have a `warn` function in addition to `print`.  We still good?
- Is the log level convention a recipe for disaster, or is it good?
- Should there be a `t.log.level` conf.lua flag that adds basic filtering to the default impl?

**Comment by bjornbytes** [comment]
Had another idea for log levels: Instead of `[INFO]`, do `@info` or `@level=info`.  This is the convention used by Lua's new `warn` function, and it kinda decouples the log metadata from how it is presented.

**Comment by bjornbytes** [comment]
This has mostly been implemented on the `log` branch.

One annoying part is that if anything fires a log message before you're able to install your logging handler, it won't get to your callback.  Overriding `lovr.log` in `conf.lua` helps a lot though.

**Comment by crertel** [comment]
Any chance there could be an option for line-delimited JSON output?

**Comment by bjornbytes** [comment]
The `lovr.log` callback could convert the message to JSON if it wanted (cjson is included).

**Comment by crertel** [comment]
Ah, awesome, thank you!

I was wondering about just a lighter `sprintf`-like solution...like just applying the correct escape character transforms and then interpolating it into a string (instead of a whole JSON serialization trip).

**Comment by bjornbytes** [comment]
This was added in 9abed05

`lovr.log` is a new callback with the following default implementation:

```lua
function lovr.log(message, level, tag)
  print(message)
end
```
